File: src/0045.py
class Solution:
    def jump(self, nums: list[int]) -> int:
        jumps = last_jump = max_idx = i = 0
        while last_jump < len(nums) - 1:
            max_idx = max(max_idx, i + nums[i])
            if i == last_jump:
                last_jump = max_idx
                jumps += 1
            i += 1
            print(i, jumps, last_jump, max_idx)
            input()
        return jumps

sample = [3,4,3,2,1,2]
Solution().jump(sample)

File: 0045.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size() - 1, i = 0;
        int jumps = 0, max_idx = 0, last_jump = 0;
        while(last_jump < n) {
            max_idx = max(max_idx, i + nums[i]);
            if(i == last_jump) {
                last_jump = max_idx;
                jumps++;
            }
            i++;
        }
        return jumps;
    }
};
